collect_logs.py

```diff
@@ -0,0 +1,164 @@
+       - nodeip: host Ip address
+       - uname: Username
+       - pword: password
+       - log_dir: directory to store all the logs
+       - results: dict: host Ip address with boolean parameters
+    """
+    print(f"Connecting {nodeip} to generate sosreport")
+    try:
+        results[nodeip] = True
+        ssh_d = SSHConnectionManager(nodeip, uname, pword).get_client()
+        ssh_d.exec_command("sudo yum -y install sos")
+        stdin, stdout, stderr = ssh_d.exec_command(
+            "sudo sosreport -a --all-logs -e ceph --batch"
+        )
+        sosreport = re.search(r"/var/tmp/sosreport-.*.tar.xz", stdout.read().decode())
+        print(f"Successfully generated sosreport in node {nodeip} :{sosreport.group()}")
+        ssh_d.exec_command(f"sudo chown {uname}:{uname} {sosreport.group()}")
+        ssh_install.exec_command(f"scp {nodeip}:{sosreport.group()} {log_dir}")
+        ssh_d.exec_command(f"sudo rm -rf {sosreport.group()}")
+        ssh_d.close()
+        print(f"Successfully copied report to {nodeip}")
+    except Exception:
+        results[nodeip] = False
+
+
+def run(args):
+    """
+    Using the provided installer ip get all other nodes then generate sosreport for all the nodes obtained.
+    Run shell script on installer node based on its ceph version, then upload all the collected logs to magna
+    :param:
+       - ip        :installer IP address
+       - username  :Username other than root
+       - password  :password for installer node
+       - log-dir   :name of the output directory [default tmp]
+    Returns:
+        0 on success or 1 for failures
+    """
+    results = dict()
+    run_id = generate_unique_id(length=6)
+    ip = args["--ip"]
+    uname = args["--username"]
+    pword = args["--password"]
+    log_dir = f"ceph_logs_{run_id}"
+    magna = "http://magna002.ceph.redhat.com/cephci-jenkins"
+
+    ssh_install = SSHConnectionManager(ip, uname, pword).get_client()
+    stdin, stdout, stderr = ssh_install.exec_command("hostname")
+    if "installer" in stdout.read().decode():
+        ssh_install.exec_command(f"sudo mkdir -p {log_dir}")
+        ssh_install.exec_command(f"sudo chown -R {uname}:{uname} {log_dir}")
+        stdin, stdout, stderr = ssh_install.exec_command(
+            "cut -f 1 /etc/hosts | cut -d ' ' -f 1"
+        )
+        nodes = stdout.read().decode().split("\n")
+        print(f"Host that are obtained from given host: {nodes}")
+        with parallel() as p:
+            for nodeip in nodes:
+                if nodeip.startswith("10.0"):
+                    p.spawn(
+                        generate_sosreport_in_node,
+                        ssh_install,
+                        nodeip,
+                        uname,
+                        pword,
+                        log_dir,
+                        results,
+                    )
+        collect_logs(ip, ssh_install, uname, log_dir, results)
+        upload_logs(log_dir, ssh_install, nodeip, results)
```

Comment:
Can do these steps in parallel?

Reply:
This operation will be performed on single node 
so parallel is not required